### DISCUSSION TITLE
Help define a profiling suite of shareware games

### DISCUSSION THREAD

**Discussion by kcgen** [dosbox-staging/dosbox-staging#1264] [General]
Help us find a minimal set of games that, collectively, exercise almost all of `dosbox-staging`'s functionality. 

This will let us add an optional profile-guided optimization (PGO) step to *every* platform that maximally optimizes the binary. The improvement can range from 3% to 15% depending on the section of code it optimizes, for an overall improvement of roughly 5%+. 

So for example, find a game from the mid '80s that supports Hercules, CGA, and EGA graphics and various sound devices like the Covox, Tandy, PC Speaker, and Adlib. 

Then find another game from the mid '90s than runs in VGA and SVGA modes, uses 486 and pentium instructions, and supports various Sound Blaster models and the Gravis Ultrasound. 

One of the catches with PGO is that we need to run the entire profiling task (all games and their various configurations) in one run of dosbox-staging. This means we're constrained to use games that are 'quitable' back to DOS using some basic key commands, like the sierra games menu system: `escape, down arrow, down arrow, enter`, which we can feed via the `autotype` command.

**Comment by dreamer** [comment]
It's not a game, but I think https://fractint.org/ will be a great test case for stressing out floating-point emulation code. Early versions of the program used integer math, but nowadays it uses floats and doubles (at least judging by the source code). It can be run in batch mode, can generate gifs that can be easily verified against reference renderings, and generating fractals in higher resolutions is CPU-intensive enough as to serve as a benchmark as well.

**Comment by kcgen** [comment]
Oh that's a great find! :+1: for batch mode that quits on its own and having verifiable output. Sounds like we can use it not only _during_ profiling but also _after_ to test core-correctness of the optimized binary.

**Comment by ronoverdrive** [comment]
There's also Phil's Computer Labs DOS benchmark kit: https://www.philscomputerlab.com/dos-benchmark-pack.html

**Comment by kcgen** [comment]
Adding FPU-related programs:
 - Carmageddon on armv7
 - MS Flight Simulator 1.0 (https://github.com/dosbox-staging/dosbox-staging/projects/7#card-67424512)

**Comment by arrowgent** [comment]
demo scene programs?

https://www.pouet.net/prodlist.php?type%5B%5D=demo&platform%5B%5D=MS-Dos

some test the limits of hardware & dos era.


similar:
aida freeware

https://archive.org/details/freesoftwarefordos_system_aida

**Comment by johnnovak** [comment]
I don't think this would do anything desirable in practice as different exercise different paths of the emulation code, hotspots will be different, etc.
